### ISSUE TITLE
1.4.X STATICFILES_FINDERS

### ISSUE THREAD

**Issue by rizumu** [jazzband/django-pipeline#378]
I was able to prove that certain combinations of STATICFILES_FINDERS will result in no html output (no stylesheets or javascript on the rendered page).

On 1.3.X I was using:

```
STATICFILES_FINDERS = [
    "pipeline.finders.FileSystemFinder",
    "pipeline.finders.AppDirectoriesFinder",
    "pipeline.finders.PipelineFinder",
    "pipeline.finders.CachedFileFinder",
]
```

After upgrading to  1.4.x and updating my templates to use the new tags `pipeline`, `stylesheet`, `javascript` I discovered the problem and found that using the following works:

```
STATICFILES_FINDERS = [
    "django.contrib.staticfiles.finders.FileSystemFinder",
    "django.contrib.staticfiles.finders.AppDirectoriesFinder",
    "pipeline.finders.PipelineFinder",
    "pipeline.finders.CachedFileFinder",
]
```

**Comment by rizumu** [comment]
I tried out the most recent version, to use non-Ruby SASS which was introduced in 1.4.3, but this issue still persists. Has anyone found a solution for yet?

**Comment by rizumu** [comment]
I just found  #328 and explicitly setting `STATICFILES_STORAGE = PIPELINE_STORAGE = "pipeline.storage.PipelineCachedStorage"` appears to fix it. I'll keep an eye on this over the coming days to be sure, but hopefully this can be resolved with some documentation and better error reporting.

**Comment by cyberdelia** [comment]
``PIPELINE_STORAGE`` isn't used, not referenced in the code anymore :expressionless:
But you effectively need to setup ``STATICFILES_STORAGE`` now.

**Comment by rizumu** [comment]
I think I may have figured it out, and might be why people who started on 1.4.x don't experience this issue. Because this was an upgrade of 1.3.x to 1.4.x, I already had rules to .gitignore the auto-generated .css files that live in the same folders as my .less files. Keeping these .css files in place somehow confused pipeline into not dealing with changes made to the less files. Now, once I remove all these auto-generated .css files, things seem to work, and no local files are generated for the .css any more. @cyberdelia does this make any sense considering what you've changed from 1.3.x to 1.4.x?

**Comment by bitcity** [comment]
@rizumu Were you able to resolve your issue with 1.4.x or 1.5.0 ? I've started using pipeline recently and I'm experiencing similar issues (https://github.com/cyberdelia/django-pipeline/issues/418)

**Comment by evdoks** [comment]
I am also experience a similar issue on v1.5

When using 
```
STATICFILES_FINDERS = (
    "pipeline.finders.FileSystemFinder",
    "pipeline.finders.AppDirectoriesFinder",
    "pipeline.finders.CachedFileFinder",
    "pipeline.finders.PipelineFinder",
)
```
all `.js` files from `bower_components` directory are ignored.
When using 
```
STATICFILES_FINDERS = (
    "django.contrib.staticfiles.finders.FileSystemFinder",
    "django.contrib.staticfiles.finders.AppDirectoriesFinder",
    "pipeline.finders.CachedFileFinder",
    "pipeline.finders.PipelineFinder",
)
```
everything works, but all bower examples, docs etc. are copied to `STATIC_ROOT`.

**Comment by samdieck** [comment]
I'm having the same issue as the OP. I started using pipeline recently and collectstatic with pipeline.finders.* will result in empty assets. If I change to django finders, it'll correctly generate the assets, however, it'll also copy the original version of files, which is what i'm trying to avoid.

**Comment by rizumu** [comment]
I haven't seen this issue since my post in March, I had considered this closed but since others are still experiencing it there might be something more going on here.

**Comment by YozhEzhi** [comment]
@rizumu +1 from 2016

**Comment by emyller** [comment]
@cyberdelia This issue is still affecting a setup with version 1.6.8.

My configuration is:

```python
STATIC_ROOT = os.path.join(BASE_DIR, 'static')

STATICFILES_DIRS = [
    os.path.join(BASE_DIR, 'assets'),
    os.path.join(BASE_DIR, 'assets', 'bower_components'),  # Shortcut
]

STATICFILES_STORAGE = 'pipeline.storage.PipelineCachedStorage'

STATICFILES_FINDERS = [
    'pipeline.finders.FileSystemFinder',
    'pipeline.finders.AppDirectoriesFinder',
    'pipeline.finders.CachedFileFinder',
    'pipeline.finders.PipelineFinder',
]

PIPELINE = {
    'JAVASCRIPT': {
        'vendor': {
            'source_filenames': [
                'underscore/underscore.js',
                'backbone/backbone.js',
                'jquery/jquery.js',
            ],
            'output_filename': 'js/vendor.js',
        },
        'modules': {
            'source_filenames': [
                'js/modules/settings.js',
                'js/modules/*.js',
            ],
            'output_filename': 'js/modules.js',
        },
    },
    'STYLESHEETS': {
        'main': {
            'source_filenames': ['styles/*.scss'],
            'output_filename': 'css/main.css',
        },
    },

    # Engine configuration
    'COMPILERS': [
        'pipeline.compilers.sass.SASSCompiler',
    ],
    'SASS_BINARY': '/usr/bin/env sassc',
    'SASS_ARGUMENTS': ' '.join([
        # Include Compass libs
        '--include-path={}'.format(os.path.join(
            BASE_DIR, 'html', 'bower_components', 'compass-mixins', 'lib')),
    ]),
}
```

No file is being found **at all**. If, though, I use `django.contrib.staticfiles.finders.FileSystemFinder` instead of `pipeline.finders.FileSystemFinder`, it does work normally.

**Comment by emyller** [comment]
Following up: I've just read the `finder.py` sourcecode and found this:

```
class FileSystemFinder(PatternFilterMixin, DjangoFileSystemFinder):
    """
    Like FileSystemFinder, but doesn't return any additional ignored patterns
    This allows us to concentrate/compress our components without dragging
    the raw versions in too.
    """
    ignore_patterns = [
        '*.js',
        '*.css',
        '*.less',
        '*.scss',
        '*.styl',
```

If `FileSystemFinder` is supposed to ignore CSS, SCSS, JavaScript, etc files, how could it ever find the files to work with? Either it is very wrong or we're all missing the point of using it.

**Comment by vskh** [comment]
Including `pipeline.finders.AppDirectoriesFinder` into STATICFILES_FINDERS breaks collectstatic for me as well. While source says that it excludes certain types of files on purpose to avoid copying them to STATIC_ROOT, I did not find any trace of logic in storage to resolve source files when compiling configured packages. As of version 1.6.11 I used for debugging it just follows the regular logic of static files storage, i.e. prepending requested files with STATIC_ROOT meaning that although it excludes files from collection it still looks for them in STATIC_ROOT :(
